libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,529 @@
+		if (util::findOffset(_operation.output, slot) || generateSlotOnTheFly(slot))
+				--previousLayoutSize;
+
+	auto layout = ranges::views::iota(0u, previousLayoutSize) |
+		ranges::views::transform([](size_t _index) { return PreviousSlot{_index}; }) |
+		ranges::to<vector<variant<PreviousSlot, StackSlot>>>;
+	// The call produces a known sequence of values.
+	layout += _operation.output;
+
+	stack = createIdealLayout(stack, layout, generateSlotOnTheFly);
+
+	if (auto const* assignment = get_if<CFG::Assignment>(&_operation.operation))
+		for (auto& stackSlot: stack)
+			if (auto const* varSlot = get_if<VariableSlot>(&stackSlot))
+				yulAssert(!util::findOffset(assignment->variables, *varSlot), "");
+
+	for (StackSlot const& input: _operation.input)
+		stack.emplace_back(input);
+
+	m_layout.operationEntryLayout[&_operation] = stack;
+
+	// Remove anything from the stack top that can be freely generated or dupped from deeper on the stack.
+	while (!stack.empty())
+	{
+		if (canBeFreelyGenerated(stack.back()))
+			stack.pop_back();
+		else if (auto offset = util::findOffset(stack | ranges::views::reverse | ranges::views::drop(1), stack.back()))
+		{
+			if (*offset + 2 < 16)
+				stack.pop_back();
+			else
+				break;
+		}
+		else
+			break;
+	}
+
+	// TODO: there may be a better criterion than overall stack size.
+	if (stack.size() > 12)
+		// Deduplicate and remove slots that can be freely generated.
+		stack = compressStack(move(stack));
+	return stack;
+}
+
+Stack StackLayoutGenerator::compressStack(Stack _stack)
+{
+	optional<size_t> firstDupOffset;
+	do
+	{
+		if (firstDupOffset)
+		{
+			if (_stack.size() - *firstDupOffset - 1 > 1)
+				std::swap(_stack.at(*firstDupOffset + 1), _stack.back());
+			std::swap(_stack.at(*firstDupOffset), _stack.back());
+			_stack.pop_back();
+			firstDupOffset.reset();
+		}
+		for (auto&& [offset, slot]: _stack | ranges::views::enumerate)
+			if (canBeFreelyGenerated(slot) || util::findOffset(_stack | ranges::views::take(offset), slot))
+				firstDupOffset = offset;
+	}
+	while (firstDupOffset);
+	return _stack;
+}
+
+Stack StackLayoutGenerator::propagateStackThroughBlock(Stack _exitStack, CFG::BasicBlock const& _block)
+{
+	Stack stack = std::move(_exitStack);
+	for (auto& operation: _block.operations | ranges::views::reverse)
+		stack = propagateStackThroughOperation(stack, operation);
+	return stack;
+}
+
+void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const& _entry)
+{
+	std::list<CFG::BasicBlock const*> toVisit{&_entry};
+	std::set<CFG::BasicBlock const*> visited;
+
+	while (!toVisit.empty())
+	{
+		// TODO: calculate backwardsJumps only once.
+		std::list<std::pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps;
+		while (!toVisit.empty())
+		{
+			CFG::BasicBlock const *block = *toVisit.begin();
+			toVisit.pop_front();
+
+			if (visited.count(block))
+				continue;
+
+			if (std::optional<Stack> exitLayout = std::visit(util::GenericVisitor{
```

Comment:
I really hate that the parameters to `std::visit` are not swapped.

Reply:
Can you extract computing the exit layout into its own function?

Reply:
We could probably add a helper, maybe even one that even integrates the ``GenericVisitor`` part, like
```
variantVisit(variant,
  [](FirstAlternative) { ... },
  [](SecondAlternative) { ... }
  ...
);
```
...